### PR TITLE
cmake: work around `ios.toolchain.cmake` breaking feature-detections

### LINKED ISSUE #15557: 8.11.0 - macOS getpass_r error building CMake
## Error on getpass

> error: call to undeclared function 'getpass_r'; ISO C99 and later do not support implicit function declarations

```
/Users/runner/work/apothecary/apothecary/apothecary/build/curl/src/tool_paramhlp.c:569:5: error: call to undeclared function 'getpass_r'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
    getpass_r(prompt, passwd, sizeof(passwd));
    ^
/Users/runner/work/apothecary/apothecary/apothecary/build/curl/src/tool_paramhlp.c:569:5: note: did you mean 'getpass'?
/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk/usr/include/unistd.h:559:7: note: 'getpass' declared here
char    *getpass(const char *) __POSIX_C_DEPRECATED(199506L);
         ^
1 error generated.
make[2]: *** [src/CMakeFiles/curl.dir/tool_paramhlp.c.o] Error 1
make[1]: *** [src/CMakeFiles/curl.dir/all] Error 2
make: *** [all] Error 2
```

```
curl/src/tool_paramhlp.c:569:5
```

Building with CMake
```
 echo "building curl $TYPE | $PLATFORM"
        echo "--------------------"
        mkdir -p "build_${TYPE}_${PLATFORM}"
        cd "build_${TYPE}_${PLATFORM}"
        rm -f CMakeCache.txt *.a *.o *.lib
        cmake  .. \
            -DCMAKE_C_STANDARD=${C_STANDARD} \
            -DCMAKE_CXX_STANDARD=${CPP_STANDARD} \
            -DCMAKE_CXX_STANDARD_REQUIRED=ON \
            -DCMAKE_CXX_FLAGS="-DUSE_PTHREADS=1 ${FLAG_RELEASE} " \
            -DCMAKE_C_FLAGS="-DUSE_PTHREADS=1 ${FLAG_RELEASE} " \
            -DCMAKE_CXX_EXTENSIONS=OFF \
            -DBUILD_SHARED_LIBS=OFF \
            -DCURL_STATICLIB=ON \
            -DBUILD_STATIC_LIBS=ON \
            -DENABLE_UNICODE=ON \
            -DCMAKE_BUILD_TYPE=Release \
            -DCMAKE_INSTALL_PREFIX=Release \
            -DDEPLOYMENT_TARGET=${MIN_SDK_VER} \
            -DCMAKE_INCLUDE_OUTPUT_DIRECTORY=include \
            -DHAVE__FSEEKI64=OFF \
            -DCMAKE_INSTALL_INCLUDEDIR=include \
            -DCMAKE_TOOLCHAIN_FILE=$APOTHECARY_DIR/toolchains/ios.toolchain.cmake \
            -DPLATFORM=$PLATFORM \
            -DENABLE_BITCODE=OFF \
            -DCMAKE_USE_SYSTEM_CURL=OFF \
            -DENABLE_ARC=ON \
            -DCMAKE_POSITION_INDEPENDENT_CODE=TRUE \
            -DCURL_DISABLE_LDAP=ON \
            -DENABLE_VISIBILITY=OFF \
            ${SSL_DEFS} \
            -DCMAKE_PREFIX_PATH="${LIBS_ROOT}" \
            -DZLIB_ROOT=${ZLIB_ROOT} \
            -DZLIB_INCLUDE_DIR=${ZLIB_INCLUDE_DIR} \
            -DZLIB_INCLUDE_DIRS=${ZLIB_INCLUDE_DIR} \
            -DZLIB_LIBRARY=${ZLIB_LIBRARY} \
            -DZLIB_LIBRARIES=${ZLIB_LIBRARY} \
            -DENABLE_ARES=OFF \
            -DCMAKE_VERBOSE_MAKEFILE=${VERBOSE_MAKEFILE} \
            -DENABLE_UNIX_SOCKETS=OFF \
            -DHAVE_LIBSOCKET=OFF \
            -DCURL_ENABLE_SSL=${CURL_ENABLE_SSL} \
            -DCMAKE_MACOSX_BUNDLE=OFF \
            -DUSE_SECURE_TRANSPORT=${USE_SECURE_TRANSPORT} \
            -DCURL_USE_SECTRANSP=${USE_SECURE_TRANSPORT} \
            -DUSE_NGHTTP2=OFF \
            -DCURL_DISABLE_POP3=ON \
            -DCURL_CA_FALLBACK=ON \
            -DCURL_DISABLE_IMAP=ON \
            -DENABLE_WEBSOCKETS=ON \
            -DENABLE_UNIX_SOCKETS=ON \
            -DCURL_BROTLI=ON \
            -DBROTLI_INCLUDE_DIRS=${LIBBROTLI_INCLUDE_DIR} \
            -DBROTLIDEC_LIBRARY=${LIBBROTLI_DEC_LIB} \
            -DBROTLICOMMON_LIBRARY=${LIBBROTLI_LIBRARY} \
            -DBROTLI_INCLUDE_DIR=${LIBBROTLI_INCLUDE_DIR} \
            -DBROTLI_LIBRARIES="${LIBBROTLI_LIBRARY} ;${LIBBROTLI_DEC_LIB};${LIBBROTLI_ENC_LIB}" \
            -DUSE_LIBIDN2=OFF \
            -DENABLE_VERBOSE=ON \
            -DENABLE_THREADED_RESOLVER=ON \
            -DENABLE_IPV6=ON
        cmake --build . --config Release --target install
        cd ..

```
[fulllog.log](https://github.com/user-attachments/files/17717178/fulllog.log)

### I expected the following

Build success

### curl/libcurl version

curl 8.11.0

### operating system

buildinfo.configure.generator: Unix Makefiles
buildinfo.configure.make: /usr/bin/make
buildinfo.host.cpu: arm64
buildinfo.host.os: Darwin
buildinfo.target.cpu: x86_64
buildinfo.target.os: Darwin
buildinfo.target.flags: APPLE UNIX CROSS
buildinfo.compiler: AppleClang
buildinfo.compiler.version: 15.0.0.15000309
buildinfo.sysroot: /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk

on macOS latest runner GitHub Actions https://github.com/actions/runner-images/blob/main/images/macos/macos-14-Readme.md

### PR DESCRIPTION
**Pull request by vszakats**
Fix builds with CMake configured to falsely return successful detection
when using `check_function_exists()` (and `check_library_exists()`, and
anything based on `try_compile()` that's relying on the linker). After
such mis-detection the build fails when trying to use the feature that
doesn't in fact exist.

The mis-detection is caused by this CMake setting:
```
set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
```
It is set by default (or on conditions) when using 3rd-party toolchain:
https://github.com/leetal/ios-cmake/blob/master/ios.toolchain.cmake

After this patch the curl build overrides this setting for the duration
of feature tests, and logs a message about it.

Also preset and skip feature tests for known mis-detections:
- `connect()` in `libsocket`
- `getpass_r()`
- `eventfd()` (did not cause an issue due to a separate bug)
- `sendmmsg()` (did not cause an issue because it's Linux-only)

If mis-detections are still seen, the workaround is to force-set the
specific feature by passing `-DHAVE_*=OFF` to cmake.
Also consider passing `-DENABLE_STRICT_TRY_COMPILE=ON` for
`ios.toolchain.cmake` to fix the root cause.

Interestingly curl itself uses this setting to speed up compile-only
detections: be17f298ff508d62d493d4a8d43e56a1e2861a50 #3744

Also:
- OtherTests.cmake: restore original value of
  `CMAKE_TRY_COMPILE_TARGET_TYPE`. Before this patch it reset it
  to empty.
- OtherTests.cmake: unset a local variable after use, quote a string.

Follow-up to 8e345057761a8f796403923a96f2c8fd3edca647 #15164
Follow-up to 8b76a8aeb21c8ae2261147af1bddd0d4637c252c #15525
Ref: https://github.com/leetal/ios-cmake/issues/47
Ref: https://gitlab.kitware.com/cmake/cmake/-/issues/18121
Ref: https://cmake.org/cmake/help/latest/variable/CMAKE_TRY_COMPILE_TARGET_TYPE.html
Reported-by: Dan Rosser
Fixes #15557

### CODE REVIEW

**Comment by vszakats** [comment]
The workaround for 8.11.0 is passing `-DHAVE_GETPASS_R=0` to cmake.

**Comment by vszakats** [comment]
Upstream issues that look similar to the fallouts reported for curl:
https://gitlab.kitware.com/cmake/cmake/-/issues/18121
https://github.com/leetal/ios-cmake/issues/47
https://github.com/leetal/ios-cmake/commit/7658382bcaf8b6c0043e25d6299761bf0dd3bb60

**Comment by vszakats** [comment]
Based on the upstream issues, this CMake call:
```cmake
  set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
```
breaks CMake's built-in `check_function_exists()` and `check_library_exists()`
feature checks by making them falsely return success.

`ios.toolchain.cmake` makes this call by default:
```cmake
[...]
# ENABLE_STRICT_TRY_COMPILE: (ON|OFF) Enables or disables strict try_compile() on all Check* directives (will run linker
#    to actually check if linking is possible). Default OFF (will set CMAKE_TRY_COMPILE_TARGET_TYPE to STATIC_LIBRARY)
[...]
if(NOT DEFINED ENABLE_STRICT_TRY_COMPILE)
  # Unless specified, disable strict try_compile()
  set(ENABLE_STRICT_TRY_COMPILE OFF)
  message(STATUS "[DEFAULTS] Using NON-strict compiler checks by default. ENABLE_STRICT_TRY_COMPILE not provided!")
endif()
set(ENABLE_STRICT_TRY_COMPILE_INT ${ENABLE_STRICT_TRY_COMPILE} CACHE BOOL
        "Whether or not to use strict compiler checks" FORCE)
[...]
if(NOT ENABLE_STRICT_TRY_COMPILE_INT)
  set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
endif()
```
https://github.com/leetal/ios-cmake/blob/ad96a372b168930c2a1ff9455e1a9ccb13021617/ios.toolchain.cmake

Earlier versions enabled this behavior unconditionally, later modified to do it for the `Xcode` generator, and later changed to the on-by-default separate option above.

TL;DR: `ios.toolchain.cmake` can break any builds using the CMake built-ins
`check_function_exists()` and `check_library_exists()`.

In report #15557, the CMake configuration already has a workaround for a
mis-detected libsocket: `-DHAVE_LIBSOCKET=OFF`. This was never seen
in CI and likely caused by the same issue.

**Comment by vszakats** [comment]
@danoli3 Does this patch resolve the issue you reported?

**Comment by danoli3** [comment]
Yes and thank you! Some insights into some other library issues with the findings in ios toolchain thanks mate!
